File: internal/categories/categories.go
// Package categories provides shared category normalization rules used by
// both the agent (Go) and the dashboard (TypeScript). The TS version in
// dashboard/lib/categories.ts MUST be kept in sync with this file.
package categories

import "strings"

// Normalized category names (English) used for aggregation + display.
const (
	Groceries     = "Groceries"
	Restaurants   = "Restaurants"
	Transport     = "Transport & Fuel"
	Insurance     = "Insurance"
	Healthcare    = "Healthcare"
	Home          = "Home & Furniture"
	Electronics   = "Electronics & Telecom"
	Fashion       = "Fashion"
	Leisure       = "Leisure"
	Travel        = "Travel"
	Municipality  = "Municipality & Gov"
	Pets          = "Pets"
	Kids          = "Kids"
	BooksMedia    = "Books & Media"
	Beauty        = "Beauty"
	Finance       = "Finance"
	Work          = "Work"
	Other         = "Other"
)

// merchantOverrides: specific merchants whose normalized category differs
// from what Cal/Max report. Match is case-insensitive substring on the
// merchant description. First match wins.
var merchantOverrides = []struct {
	match      string // lowercase substring
	normalized string
}{
	{"wolt", Groceries},
	{"האסישוק", Groceries},
	{"hacishuq", Groceries},
	{"משק קירשנר", Groceries},
	{"kirshner", Groceries},
	{"גליקסמן", Groceries},

	{"קיי אס פי", Home},
	{"ksp", Home},
	{"המשרדיה", Home},
	{"סטוק סנטר", Home},

	{"בוימן", Pets},
	{"boyman", Pets},
	{"זו סנטר", Pets},
	{"מיץ פטל פטס", Pets},
	{"יובל סמואל", Pets},

	{"סול מרכזי", Kids},
	{"גאחלנד", Kids},
	{"גיילנד", Kids},
	{"gayaland", Kids},
	{"פינת חי", Kids},
	{"אלמוג פנאי ומוזיקה", Kids},
	{"יולי אנד איב", Kids},
	{"הפיראט האדום", Kids},
	{"טויס", Kids},
	{"toys", Kids},

	{"upapp", Leisure},
	{"audible", Leisure},

	{"איבן בייקרי", Restaurants},
	{"ibn bakery", Restaurants},

	{"פחות מאלף", Other},
	{"pchot", Other},
	{"רד בוקס", Other},
	{"red box", Other},

	{"vectify", Work},
	{"sqsp", Work},
	{"squarespace", Work},
	{"anthropic", Work},

	{"פנגו", Transport},
	{"pango", Transport},
	{"חניון", Transport},

	{"אמישרגז", Municipality},
	{"חברת החשמל", Municipality},

	{"ארנק מט", Finance},
	{"מכירת מט", Finance},
}

// rawCategoryMap maps provider-supplied category strings (Cal + Max, Hebrew)
// to normalized English names. Keys are compared as exact-match after trim.
var rawCategoryMap = map[string]string{
	// Groceries
	"מזון ומשקאות":   Groceries,
	"מזון וצריכה":    Groceries,
	"מזון מהיר":      Groceries,

	// Restaurants
	"מסעדות":             Restaurants,
	"מסעדות, קפה וברים":  Restaurants,

	// Transport & Fuel
	"רכב ותחבורה":   Transport,
	"תחבורה ורכבים": Transport,
	"אנרגיה":        Transport,
	"דלק, חשמל וגז":  Transport,

	// Insurance
	"ביטוח":          Insurance,
	"ביטוח ופיננסים": Insurance,

	// Healthcare
	"רפואה ובריאות":     Healthcare,
	"רפואה ובתי מרקחת":  Healthcare,

	// Home & Furniture
	"ריהוט ובית":  Home,
	"עיצוב הבית":  Home,

	// Electronics & Telecom
	"תקשורת ומחשבים": Electronics,
	"חשמל ומחשבים":   Electronics,
	"שירותי תקשורת":   Electronics,

	// Fashion
	"אופנה": Fashion,

	// Leisure
	"פנאי בילוי":          Leisure,
	"פנאי, בידור וספורט":  Leisure,

	// Travel
	"תיירות":          Travel,
	"מלונאות ואירוח":  Travel,
	"טיסות ותיירות":   Travel,

	// Municipality & Gov
	"מוסדות":        Municipality,
	"עירייה וממשלה": Municipality,

	// Pets
	"חיות מחמד": Pets,

	// Kids
	"ילדים": Kids,

	// Books & Media
	"ספרים ודפוס": BooksMedia,

	// Beauty
	"טיפוח ויופי":     Beauty,
	"קוסמטיקה וטיפוח": Beauty,

	// Finance
	"פיננסים":     Finance,
	"העברת כספים": Finance,
	"משיכת מזומן": Finance,

	// Office/equipment — falls through to Other (rare in this dataset).
	"ציוד ומשרד": Other,
}

// Normalize returns the canonical English category for a given merchant +
// raw category. Merchant overrides take precedence over category mapping.
// Unknown inputs fall through to "Other".
func Normalize(merchant, rawCategory string) string {
	merchantLower := strings.ToLower(strings.TrimSpace(merchant))
	for _, rule := range merchantOverrides {
		if strings.Contains(merchantLower, rule.match) {
			return rule.normalized
		}
	}
	if norm, ok := rawCategoryMap[strings.TrimSpace(rawCategory)]; ok {
		return norm
	}
	return Other
}

File: dashboard/lib/categories.ts
// Category normalization rules. This file MUST be kept in sync with
// internal/categories/categories.go in the Go module. If you change one,
// change the other.

export const Category = {
  Groceries: "Groceries",
  Restaurants: "Restaurants",
  Transport: "Transport & Fuel",
  Insurance: "Insurance",
  Healthcare: "Healthcare",
  Home: "Home & Furniture",
  Electronics: "Electronics & Telecom",
  Fashion: "Fashion",
  Leisure: "Leisure",
  Travel: "Travel",
  Municipality: "Municipality & Gov",
  Pets: "Pets",
  Kids: "Kids",
  BooksMedia: "Books & Media",
  Beauty: "Beauty",
  Finance: "Finance",
  Work: "Work",
  Other: "Other",
} as const;

// Merchant-level overrides (checked first). Case-insensitive substring match
// on merchant description.
const merchantOverrides: Array<{ match: string; normalized: string }> = [
  { match: "wolt", normalized: Category.Groceries },
  { match: "האסישוק", normalized: Category.Groceries },
  { match: "hacishuq", normalized: Category.Groceries },
  { match: "משק קירשנר", normalized: Category.Groceries },
  { match: "kirshner", normalized: Category.Groceries },
  { match: "גליקסמן", normalized: Category.Groceries },

  { match: "קיי אס פי", normalized: Category.Home },
  { match: "ksp", normalized: Category.Home },
  { match: "המשרדיה", normalized: Category.Home },
  { match: "סטוק סנטר", normalized: Category.Home },

  { match: "בוימן", normalized: Category.Pets },
  { match: "boyman", normalized: Category.Pets },
  { match: "זו סנטר", normalized: Category.Pets },
  { match: "מיץ פטל פטס", normalized: Category.Pets },

  { match: "סול מרכזי", normalized: Category.Kids },
  { match: "גאחלנד", normalized: Category.Kids },
  { match: "גיילנד", normalized: Category.Kids },
  { match: "gayaland", normalized: Category.Kids },
  { match: "פינת חי", normalized: Category.Kids },
  { match: "אלמוג פנאי ומוזיקה", normalized: Category.Kids },
  { match: "יולי אנד איב", normalized: Category.Kids },

  { match: "upapp", normalized: Category.Leisure },
  { match: "audible", normalized: Category.Leisure },

  { match: "איבן בייקרי", normalized: Category.Restaurants },
  { match: "ibn bakery", normalized: Category.Restaurants },

  { match: "פחות מאלף", normalized: Category.Other },
  { match: "pchot", normalized: Category.Other },

  { match: "vectify", normalized: Category.Work },
  { match: "sqsp", normalized: Category.Work },
  { match: "squarespace", normalized: Category.Work },
  { match: "anthropic", normalized: Category.Work },

  { match: "פנגו", normalized: Category.Transport },
  { match: "pango", normalized: Category.Transport },

  { match: "אמישרגז", normalized: Category.Municipality },
];

// Provider-supplied category → normalized English category.
// Exact match after trim.
const rawCategoryMap: Record<string, string> = {
  // Groceries
  "מזון ומשקאות": Category.Groceries,
  "מזון וצריכה": Category.Groceries,
  "מזון מהיר": Category.Groceries,
  // Restaurants
  "מסעדות": Category.Restaurants,
  "מסעדות, קפה וברים": Category.Restaurants,
  // Transport & Fuel
  "רכב ותחבורה": Category.Transport,
  "תחבורה ורכבים": Category.Transport,
  "אנרגיה": Category.Transport,
  "דלק, חשמל וגז": Category.Transport,
  // Insurance
  "ביטוח": Category.Insurance,
  "ביטוח ופיננסים": Category.Insurance,
  // Healthcare
  "רפואה ובריאות": Category.Healthcare,
  "רפואה ובתי מרקחת": Category.Healthcare,
  // Home & Furniture
  "ריהוט ובית": Category.Home,
  "עיצוב הבית": Category.Home,
  // Electronics & Telecom
  "תקשורת ומחשבים": Category.Electronics,
  "חשמל ומחשבים": Category.Electronics,
  "שירותי תקשורת": Category.Electronics,
  // Fashion
  "אופנה": Category.Fashion,
  // Leisure
  "פנאי בילוי": Category.Leisure,
  "פנאי, בידור וספורט": Category.Leisure,
  // Travel
  "תיירות": Category.Travel,
  "מלונאות ואירוח": Category.Travel,
  "טיסות ותיירות": Category.Travel,
  // Municipality & Gov
  "מוסדות": Category.Municipality,
  "עירייה וממשלה": Category.Municipality,
  // Pets
  "חיות מחמד": Category.Pets,
  // Kids
  "ילדים": Category.Kids,
  // Books & Media
  "ספרים ודפוס": Category.BooksMedia,
  // Beauty
  "טיפוח ויופי": Category.Beauty,
  "קוסמטיקה וטיפוח": Category.Beauty,
  // Finance
  "פיננסים": Category.Finance,
  "ציוד ומשרד": Category.Finance,
  "העברת כספים": Category.Finance,
  "משיכת מזומן": Category.Finance,
};

/**
 * Returns the canonical English category for a given merchant + raw category.
 * Merchant overrides take precedence. Unknown inputs fall through to "Other".
 */
export function normalizeCategory(
  merchant: string,
  rawCategory: string,
): string {
  const m = merchant.toLowerCase().trim();
  for (const rule of merchantOverrides) {
    if (m.includes(rule.match)) {
      return rule.normalized;
    }
  }
  return rawCategoryMap[rawCategory.trim()] ?? Category.Other;
}
